tests/e2e/vfile_demote_promote_test.go

```diff
@@ -0,0 +1,183 @@
+// Copyright 2017 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This test suite includes test cases to verify basic functionality
+// before upgrade for upgrade test
+
+// +build runoncevfile
+
+package e2e
+
+import (
+	"log"
+	"strconv"
+	"time"
+
+	"github.com/vmware/docker-volume-vsphere/tests/utils/dockercli"
+	"github.com/vmware/docker-volume-vsphere/tests/utils/inputparams"
+	"github.com/vmware/docker-volume-vsphere/tests/utils/misc"
+	"github.com/vmware/docker-volume-vsphere/tests/utils/verification"
+	. "gopkg.in/check.v1"
+)
+
+// vFile Demote/Promote test:
+// This test is used to check vFile volume functionality after swarm node role change.
+// In this test, we first create a vFile volume before the role change.
+// Then a worker node is promoted to manager, and the original manager is demoted to worker.
+// After the role change, we create another vFile volume.
+// At last, we attach to both of the two vFile volumes to verify their functionality.
+
+type VFileDemotePromoteTestSuite struct {
+	config         *inputparams.TestConfig
+	esx            string
+	master         string
+	worker1        string
+	volName1       string
+	volName2       string
+	container1Name string
+	container2Name string
+}
+
+func (s *VFileDemotePromoteTestSuite) SetUpSuite(c *C) {
+	s.config = inputparams.GetTestConfig()
+	if s.config == nil {
+		c.Skip("Unable to retrieve test config, skipping basic vfile tests")
+	}
+
+	s.esx = s.config.EsxHost
+	s.master = inputparams.GetSwarmManager1()
+	s.worker1 = inputparams.GetSwarmWorker1()
+}
+
+func (s *VFileDemotePromoteTestSuite) SetUpTest(c *C) {
+	s.volName1 = inputparams.GetVFileVolumeName()
+	s.volName2 = inputparams.GetVFileVolumeName()
+	s.container1Name = inputparams.GetUniqueContainerName(c.TestName())
+	s.container2Name = inputparams.GetUniqueContainerName(c.TestName())
+}
+
+var _ = Suite(&VFileDemotePromoteTestSuite{})
+
+// All VMs are created in a shared datastore
```

Comment:
Test overall looks good. Can we add some enhancement?
1. before promote/demote, after create and attach the 1st volume, write some data in the volume
2. after promote/demote, read the data back from the 1st volume to make sure data written in step 1 are still exist

Reply:
The read/write tests have been covered by the advanced_vfile_test.
Also the role change code only affects ETCD, and won't affect neither the file server nor the internal volumes.
So I think this one should focus on the role change only?